File: stree_cpp/source_files/CSourceFileID.cpp
#include "CSourceFileID.h"

namespace source_files
{
   long CSourceFileID::m_SourceFileIDGenerator = CSourceFileID::DEFAULT_SOURCE_FILE_ID;
   std::stack<long> CSourceFileID::m_MissingSourceFileIDs;
   
   long CSourceFileID::GetUniqueID()
   {
      if(!m_MissingSourceFileIDs.empty())
      {
         long id = m_MissingSourceFileIDs.top();
         m_MissingSourceFileIDs.pop();
         return id;
      }
      return ++m_SourceFileIDGenerator;
   }

   void CSourceFileID::RestoreUniqueID(long FileID)
   {
      if(FileID == m_SourceFileIDGenerator)
         --m_SourceFileIDGenerator;
      else
         m_MissingSourceFileIDs.push(FileID);
   }

   CSourceFileID::CSourceFileID(const std::string& fileName)
      : m_fileName(fileName), m_SourceFileID(GetUniqueID())
   {}

   CSourceFileID::~CSourceFileID()
   {
      RestoreUniqueID(m_SourceFileID);
   }

   long CSourceFileID::GetSourceFileID() const
   {
      return m_SourceFileID;
   }

   std::string CSourceFileID::GetSourceFileName() const
   {
      return m_fileName;
   }

   CSourceFileID::Ptr CSourceFileID::Generate(const std::string& fileName)
   {
      CSourceFileID::Ptr obj = new CSourceFileID(fileName);
      return obj;
   }

   bool CSourceFileID::Equals(const CSourceFileID& other) const
   {
      bool result = (m_SourceFileID == other.m_SourceFileID && m_fileName == other.m_fileName);
      return result;
   }

   int CSourceFileID::CompareTo(const CSourceFileID& other) const
   {
      if (m_SourceFileID == other.m_SourceFileID)
      {
         return m_fileName.compare(other.m_fileName);
      }
      return (m_SourceFileID > other.m_SourceFileID) ? 1 : -1;
   }
}

File: iSourceFilesLibrary/Classes/SourceFile/CSourceFileID.cs
using System.Collections.Generic;
using System.Threading;
using ISourceFilesLibrary.Classes.CodeUnit;

namespace ISourceFilesLibrary.Classes.SourceFile
{
   /// <summary>
   /// 
   /// </summary>
   public sealed class CSourceFileID
   {
      #region // Статические поля и функции класса

      /// <summary>
      /// Генератор уникальных номеров
      /// </summary>
      private static long m_SourceFileIDGenerator = CExtendedCodeUnit.DEFAULT_SOURCE_FILE_ID;
      /// <summary>
      /// Пропущенные (освобождённые) уникальные номера файлов
      /// </summary>
      private static Stack<long> m_MissingSourceFileIDs = new Stack<long>();

      /// <summary>
      /// Получить уникальный идентификатор файла
      /// </summary>
      /// <returns></returns>
      private static long GetUniqueID()
      {
         if (m_MissingSourceFileIDs.Count > 0)
         {
            lock (m_MissingSourceFileIDs)
            {
               return m_MissingSourceFileIDs.Pop();
            }
         }

         return Interlocked.Increment(ref m_SourceFileIDGenerator);
      }

      /// <summary>
      /// Освободить уникальный номер файла для повторного использования
      /// </summary>
      /// <param name="FileID"></param>
      private static void RestoreUniqueID(long FileID)
      {
         if (FileID == m_SourceFileIDGenerator)
         {
            System.Threading.Interlocked.Decrement(ref m_SourceFileIDGenerator);
         }
         else
         {
            lock (m_MissingSourceFileIDs)
            {
               m_MissingSourceFileIDs.Push(FileID);
            }
         }
      }

      #endregion

      #region // Члены класса

      /// <summary>
      /// Уникальный идентификатор файла
      /// </summary>
      private long m_SourceFileID;

      #endregion

      /// <summary>
      /// 
      /// </summary>
      public CSourceFileID()
      {
         m_SourceFileID = GetUniqueID();
      }

      #region // Деструктор

      /// <summary>
      /// Пользовательский деструктор
      /// </summary>
      ~CSourceFileID()
      {
         RestoreUniqueID(m_SourceFileID);
      }

      #endregion

      /// <summary>
      /// Уникальный идентификатор файла
      /// </summary>
      public long SourceFileID
      {
         get
         {
            return m_SourceFileID;
         }
      }
   };
}